### ISSUE TITLE
results are not stable | very slow

### ISSUE THREAD

**Issue by vcjob** [espnet/espnet#1686]
Hello everyone!

I am using one of the ASR egs. When I try to decide which beam-size, ctc-weight and lm-weight should I use, but while repeating inference with the same parameters, results are different each time!
This is how the model is being loaded:
```
from espnet.asr.pytorch_backend.asr import load_trained_model
model, train_args = load_trained_model(model_dir)
model.to('cuda')                                                                                                     
model.eval()

```
How to make it return the same results each time? I use beam search with size 2 and always get the 1st result out of it.
-------
Another question I have is, why it is so slow on wavs longer ~15 secs? What can I do about it? I mean, `encoded = model.encode(torch.as_tensor(fbank).cuda())` is really slow, especially on CPU.

Thank you a lot!

**Comment by sw005320** [comment]
> Hello everyone!
> 
> I am using one of the ASR egs. When I try to decide which beam-size, ctc-weight and lm-weight should I use, but while repeating inference with the same parameters, results are different each time!
> This is how the model is being loaded:
> 
> ```
> from espnet.asr.pytorch_backend.asr import load_trained_model
> model, train_args = load_trained_model(model_dir)
> model.to('cuda')                                                                                                     
> model.eval()
> ```
>
> How to make it return the same results each time? I use beam search with size 2 and always get the 1st result out of it.

Thanks for the report.
Could you let us know more details?
We might make some part non-deterministic (we try not to do it as much as possible).

> Another question I have is, why it is so slow on wavs longer ~15 secs? What can I do about it? I mean, `encoded = model.encode(torch.as_tensor(fbank).cuda())` is really slow, especially on CPU.

We have several issues in the decoder but I have never had serious issues in the encoder. Could you let us know how long it would take for the encoder calculations for ~15 sec utterances?

**Comment by vcjob** [comment]
> > Hello everyone!
> > I am using one of the ASR egs. When I try to decide which beam-size, ctc-weight and lm-weight should I use, but while repeating inference with the same parameters, results are different each time!
> > This is how the model is being loaded:
> > ```
> > from espnet.asr.pytorch_backend.asr import load_trained_model
> > model, train_args = load_trained_model(model_dir)
> > model.to('cuda')                                                                                                     
> > model.eval()
> > ```
> > 
> > 
> > How to make it return the same results each time? I use beam search with size 2 and always get the 1st result out of it.
> 
> Thanks for the report.
> Could you let us know more details?
> We might make some part non-deterministic (we try not to do it as much as possible).
> 
> > Another question I have is, why it is so slow on wavs longer ~15 secs? What can I do about it? I mean, `encoded = model.encode(torch.as_tensor(fbank).cuda())` is really slow, especially on CPU.
> 
> We have several issues in the decoder but I have never had serious issues in the encoder. Could you let us know how long it would take for the encoder calculations for ~15 sec utterances?

Thank you for the reply!
Regarding the inference time - apparently it seems to be RAM issue - it uses almost all available ram on long wavs, so Linux uses slow swap.
Now, this is longer script I used to get ASR results:

```
    with open(model_dir + "exp/train_pytorch_train/results/model.json", "r") as f:
        idim, odim, conf = json.load(f)
    model, train_args = load_trained_model(model_dir + "exp/train_pytorch_train/results/model.val10.avg.best")
    model.to('cpu')                                                                                                   $
    model.eval()
    vocab = conf["char_list"]
    subp.call(['bash','compute_feats.sh',wav_file])
    c = time.time()-a
    root = "asr1"
    with open(path + "decode/"+wav_fn.replace('.wav','')+"/dump/data.json", "r") as f:
        test_json = json.load(f)["utts"]
    key, info = list(test_json.items())[0]
    fbank = kaldiio.load_mat(info["input"][0]["feat"])
    with torch.no_grad():
        res = model.recognize(torch.as_tensor(fbank), train_args)
#        print (res)
        res = ("".join(vocab[y] for y in res[0]['yseq'])).replace("<space>", " ").replace("<eos>", "").replace('▁', ' ')
```
The results I get changes from time to time with the same file.
The parameters I have on decode.yaml are:
```
beam-size: 2
ctc-weight: 0.2
lm-weight: 0.1
maxlenratio: 0.0
minlenratio: 0.0
penalty: 0.0
```
Thank you!

**Comment by sw005320** [comment]
> > > Hello everyone!
> > > I am using one of the ASR egs. When I try to decide which beam-size, ctc-weight and lm-weight should I use, but while repeating inference with the same parameters, results are different each time!
> > > This is how the model is being loaded:
> > > ```
> > > from espnet.asr.pytorch_backend.asr import load_trained_model
> > > model, train_args = load_trained_model(model_dir)
> > > model.to('cuda')                                                                                                     
> > > model.eval()
> > > ```
> > > 
> > > 
> > > How to make it return the same results each time? I use beam search with size 2 and always get the 1st result out of it.
> > 
> > 
> > Thanks for the report.
> > Could you let us know more details?
> > We might make some part non-deterministic (we try not to do it as much as possible).
> > > Another question I have is, why it is so slow on wavs longer ~15 secs? What can I do about it? I mean, `encoded = model.encode(torch.as_tensor(fbank).cuda())` is really slow, especially on CPU.
> > 
> > 
> > We have several issues in the decoder but I have never had serious issues in the encoder. Could you let us know how long it would take for the encoder calculations for ~15 sec utterances?
> 
> Thank you for the reply!
> Regarding the inference time - apparently it seems to be RAM issue - it uses almost all available ram on long wavs, so Linux uses slow swap. On files like 1 min length it takes around 1 minute to get the results on I7-7700hq

We usually assume ~10 seconds of the utterance. 1 min length is too large and it's better to chunk the audio by using VAD or whatever.

> Now regarding the second issue. Here below is longer script I used to get ASR results:
> 
> ```
>     with open(model_dir + "exp/train_pytorch_train/results/model.json", "r") as f:
>         idim, odim, conf = json.load(f)
>     model, train_args = load_trained_model(model_dir + "exp/train_pytorch_train/results/model.val10.avg.best")
>     model.to('cpu')                                                                                                   $
>     model.eval()
>     vocab = conf["char_list"]
>     subp.call(['bash','compute_feats.sh',wav_file])
>     root = "asr1"
>     with open(path + "decode/"+wav_fn.replace('.wav','')+"/dump/data.json", "r") as f:
>         test_json = json.load(f)["utts"]
>     key, info = list(test_json.items())[0]
>     fbank = kaldiio.load_mat(info["input"][0]["feat"])
>     with torch.no_grad():
>         res = model.recognize(torch.as_tensor(fbank), train_args)
>         res = ("".join(vocab[y] for y in res[0]['yseq'])).replace("<space>", " ").replace("<eos>", "").replace('▁', ' ')
> ```
> 
> The results (res) I get changes from time to time with the same file. I mean, they differ a bit, like multiple words. Seems like variants in beam search has different scores each run.
> The parameters I have on decode.yaml are:
> 
> ```
> beam-size: 2
> ctc-weight: 0.2
> lm-weight: 0.1
> maxlenratio: 0.0
> minlenratio: 0.0
> penalty: 0.0
> ```
> 
> Thank you!

Did you fix the seed appropriately and make the operation deterministic?
We need such treatment (e.g., all our binaries call https://github.com/espnet/espnet/blob/master/espnet/utils/deterministic_utils.py)
